atf-sh/atf-check.cpp

```diff
@@ -119,15 +119,14 @@ class temp_file : public std::ostream {
             atf::env::get("TMPDIR", "/tmp")) / pattern;
 
         std::string file_s = file.str();
-        std::vector<char> buf(file_s.begin(), file_s.end() + 1);
 
-        m_fd = ::mkstemp(buf.data());
+        m_fd = ::mkstemp(&file_s[0]);
```

Comment:
As for me, the `file_s.c_str()` communicates the intention better, calling libc is exactly one of the cases it is meant for. And we could go further and use `atf::fs::path::c_str()` directly w/o the extra `file_s` variable, i.e. `&file_s[0]` -> `file.c_str()`. 

I understand that currently proposed code tries to hide `const` and make compilers happy, but actually we should talk about another question here. The `mkstemp` writes to the provided location, and AFAIR `c_str()` and `data()` work the same way and writing to the location they provide is undefined behavior. Probably, that's why we had a correctly writable vector of chars before. As long as the project is expected to be cross-platform, i.e. it should be ready for quite wide range of arch/OS/compiler combinations, probably we should take an extra care here even in exchange for an extra buffer used. With the idea to eventually change it to use C++ way in the latest standard and/or boost lib.

Anyway, I would come back to the top level question: "What are we fixing here?".

Reply:
> and AFAIR c_str() and data() work the same way and writing to the location they provide is undefined behavior.

No, they don't. At least not in C++17, where a non-const overload is provided, which allows modifying till pointer + total_size - 1. [string.accessors](http://eel.is/c++draft/basic.string#string.accessors-6)

Reply:
> > and AFAIR c_str() and data() work the same way and writing to the location they provide is undefined behavior.
> 
> No, they don't. At least not in C++17, where a non-const overload is provided, which allows modifying till pointer + total_size - 1. [string.accessors](http://eel.is/c++draft/basic.string#string.accessors-6)

Yes, it would follows this: 

> With the idea to eventually change it to use C++ way in the latest standard

And as I see, the project is currently migrating to C++14. So, probably we should not change the working code to a potential UB until the minimum requirement is C++17.

Let me share my context of the reasoning. In this specific case I am less concerned with the eventual code behind the function. I am more worried about the time of developers working on other projects (e.g. FreeBSD), which is already in short supply for analyzing failed tests, and it's frustrating to spend it on infrastructure issues rather than actual functional bugs.

Yes, it feels like the change should work fine for frequently used arch/OS/compiler combinations, but is it a useful change if it brings no benefit and only potential risk? Don't take me wrong, that's just a discussion to think out loud and compare options aimed at finding the best for the project.

Reply:
Well, yeah, that's a valid point. It's probably better to keep what's working for now.

Reply:
I really appreciate the discussion.
The code predated C++11, so some of the choices and constructs in place don’t make as much sense in a modern C++ world.
I am working towards getting all of the projects to C++14, since it’s a low barrier of entry (the projects has been built on some more esoteric platforms, like IllumOS), but note that the projects are moving towards C++17 in the near future with better deprecation messaging.
I’ll add a “chickenswitch” for C++14 vs C++17 and add a TODO comment to remove C++14 support.
The projects are all dependent on one another (kyua depends on lutok and atf and lutok depends on atf), so any major change like this needs to be done from the top down. Kyua’s last release was almost a decade ago, so I’m doing what I can to get another solid release out ASAP, since there are a ton of changes queued up on the mainline development branch.

Reply:
Also, if it were up to me, I’d mark this project “deprecated” and suggest folks use googletest instead of ATF (at the very least for C++ testing) for many of the reasons @ihoro cites, but I don’t really have the time or energy to lead rewriting all of the ATF tests from FreeBSD and NetBSD.. it was enough fun converting a ton of them from “plain” to “TAP”/“ATF” and addressing determinism problems associated with them. I don't see the value in deckchair shuffling right now.